pkg/ca/intermediateca/intermediateca.go

```diff
@@ -0,0 +1,215 @@
+
+type intermediateCA struct {
+	sync.RWMutex
+
+	// certs is a chain of certificates from intermediate to root
+	certs  []*x509.Certificate
+	signer crypto.Signer
+
+	// GCP CA Service
+	parent string
+	client *privateca.CertificateAuthorityClient
+
+	updatedCerts chan []*x509.Certificate
+}
+
+func fetchCACertificate(ctx context.Context, parent string, client *privateca.CertificateAuthorityClient, signer crypto.Signer) ([]*x509.Certificate, error) {
+	pemPubKey, err := cryptoutils.MarshalPublicKeyToPEM(signer.Public())
+	if err != nil {
+		return nil, err
+	}
+
+	isCa := true
+	// default value of 0 for int32
+	var maxIssuerPathLength int32
+
+	csr := &privatecapb.CreateCertificateRequest{
+		Parent: parent,
+		Certificate: &privatecapb.Certificate{
+			// Two week lifetime for CA certificate
+			Lifetime: durationpb.New(time.Hour * 24 * 14),
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					PublicKey: &privatecapb.PublicKey{
+						Format: privatecapb.PublicKey_PEM,
+						Key:    pemPubKey,
+					},
+					X509Config: &privatecapb.X509Parameters{
+						KeyUsage: &privatecapb.KeyUsage{
+							BaseKeyUsage: &privatecapb.KeyUsage_KeyUsageOptions{
+								CertSign: true,
+								CrlSign:  true,
+							},
+							ExtendedKeyUsage: &privatecapb.KeyUsage_ExtendedKeyUsageOptions{
+								CodeSigning: true,
+							},
+						},
+						CaOptions: &privatecapb.X509Parameters_CaOptions{
+							IsCa:                &isCa,
+							MaxIssuerPathLength: &maxIssuerPathLength,
+						},
+					},
+					SubjectConfig: &privatecapb.CertificateConfig_SubjectConfig{
+						Subject: &privatecapb.Subject{
+							CommonName:   "sigstore-intermediate",
+							Organization: "sigstore.dev",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	resp, err := client.CreateCertificate(ctx, csr)
+	if err != nil {
+		return nil, err
+	}
+
+	var pemCerts []string
+	pemCerts = append(pemCerts, resp.PemCertificate)
+	pemCerts = append(pemCerts, resp.PemCertificateChain...)
+
+	var parsedCerts []*x509.Certificate
+	for _, c := range pemCerts {
+		certs, err := cryptoutils.UnmarshalCertificatesFromPEM([]byte(c))
+		if err != nil {
+			return nil, err
+		}
+		if len(certs) != 1 {
+			return nil, errors.New("unexpected number of certificates returned")
+		}
+		parsedCerts = append(parsedCerts, certs[0])
+	}
+	log.Logger.Info("Current CA certificate chain:")
+	log.Logger.Info(pemCerts)
+
+	return parsedCerts, nil
+}
+
+func refreshCACertificate(ctx context.Context, ica *intermediateCA) {
+	ticker := time.NewTicker(time.Hour * 4)
+	for range ticker.C {
+		ica.RLock()
+		currentCert := ica.certs[0]
+		ica.RUnlock()
+
+		// Refresh certificate 7 days before expiration
+		if time.Until(currentCert.NotAfter) < (time.Hour * 24 * 7) {
+			certs, err := fetchCACertificate(ctx, ica.parent, ica.client, ica.signer)
+			if err != nil {
+				// An intermittent error is acceptable
+				log.Logger.Error(err)
```

Comment:
make this log message more informative since it would "randomly" appear without any request context